### PR TITLE
Bluetooth: L2CAP: Document behavior of alloc_buf

### LINKED ISSUE #28384: Bluetooth: L2CAP: Bad CoC SDU segment handling
**Bug description**
This is a bug happening when handling a fragmented SDU, and when not using a custom allocator function/memory pool for CoC.

When l2cap_chan_send() is called with a netbuf allocated from a pool with its buffer size equal to the exact data payload length, l2cap_chan_send() will segment the SDU and create an additional 2-byte PDU (in order to send the sdu_len).

On the receiving side, if there is not a custom allocator function defined (.alloc callback in chan_ops):

- l2cap will not mark the first PDU as a segment in a longer SDU
- on the second PDU (SDU's last 2 bytes), l2cap (in the same function) will try to read the sdu_len field, which will be garbage (or zero), and will disconnect the l2cap CoC channel.

In this case I got garbage for the second sdu_len read in my private repo, but could only reproduce getting zero when modifying a Zephyr sample.

**To Reproduce**
Check out https://github.com/jori-nordic/zephyr/commits/sdu_len_rx_bug

- Run bluetooth central sample on one device (I used nRF52840 DK)
- Run bluetooth peripheral_hr on another device
- Check the peripheral logs for warnings with sdu_len=0. e.g.
`[00:00:16.904,479] [1;33m<wrn> bt_l2cap: chan 0x20000890 len 15 sdu_len 0[0m`

To trigger the bug in another way, set TRIGGER_SDU_BUG_AGAIN to 0 and TRIGGER_SDU_BUG to 1 in samples/bluetooth/central/src/main.c 

**Expected behavior**
Correct fragmented SDU handling, i.e. not treating the second PDU as its own SDU, and not re-reading the sdu_len field.

**Impact**
Annoyance.

**Logs and console output**
[central.log.1.txt](https://github.com/zephyrproject-rtos/zephyr/files/5222922/central.log.1.txt)
[periph.log.1.txt](https://github.com/zephyrproject-rtos/zephyr/files/5222924/periph.log.1.txt)

[central.log.txt](https://github.com/zephyrproject-rtos/zephyr/files/5222923/central.log.txt)
[periph.log.txt](https://github.com/zephyrproject-rtos/zephyr/files/5222926/periph.log.txt)

**Environment (please complete the following information):**
 - OS: Linux
 - Toolchain: Zephyr SDK
 - Parent commit SHA: c64e0ce923cde54415d203e447a1dc62894ed2b0

### PR DESCRIPTION
**Pull request by Vudentz**
Channel that requires segmentation must set alloc_buf in order for the
stack to be able to reassemble segments of an SDU.

If this happens at runtime warn the user and truncate the MTU to
force segmentation to be disabled.

Fixed #28384

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

### CODE REVIEW

File: subsys/bluetooth/host/l2cap.c [hunk 1]
```diff
@@ -778,6 +778,19 @@ static void l2cap_chan_rx_init(struct bt_l2cap_le_chan *chan)
 		chan->rx.mtu = L2CAP_MAX_LE_MTU;
 	}
 
+	/* MPS shall not be bigger than MTU + 2 as the remaining bytes cannot
+	 * be used.
+	 */
+	chan->rx.mps = MIN(chan->rx.mtu + 2, L2CAP_MAX_LE_MPS);
+
+	/* Assert if channel have disabled segmentation but still have set an
```
Comment (joerchan) [inline subsys/bluetooth/host/l2cap.c:786]: The comment does not match the code, needs to be updated.

**Review by joerchan — APPROVED** [review]

**Review by jhedberg — CHANGES_REQUESTED** [review]
The changes look fine, but the commit message summary is rather misleading: when I read it I assume that there are only documentation changes involved, however when I look at the actual changes I see you're changing code as well. Could you change the commit message to something more accurate?

**Comment by jhedberg** [comment]
..or alternatively split this into two commits: one that updates alloc_buf documentation and another which does the code change

**Comment by Vudentz** [comment]
> ..or alternatively split this into two commits: one that updates alloc_buf documentation and another which does the code change

Done.

**Review by jhedberg — APPROVED** [review]

**Pull request merged**